### PR TITLE
Implement uploading output of failed tests

### PR DESCRIPTION
**Pull request by aradi**
Currently only for the GNU compiler.

### CODE REVIEW

**Comment by aradi** [comment]
Note, documentation of the new script is not correct, it must be still updated to the latest changes I've made before merging...

**Comment by aradi** [comment]
Now, it should be OK.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -130,9 +130,21 @@ jobs:
         else
           echo "TEST_OMP_PROCS=2" >> $GITHUB_ENV
         fi
-        pushd ${BUILD_DIR}
-        ctest -j ${TEST_OMP_PROCS} --output-on-failure
-        popd
+        echo "TEST_UUID=$(uuidgen)" >> ${GITHUB_ENV}
+        ctest -j 2 --output-on-failure --test-dir ${BUILD_DIR}
+
+    - name: Archive failed test output
```
Comment (bhourahine) [inline .github/workflows/build.yml:136]:
```suggestion
    - name: Archive any failed test output
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -130,9 +130,21 @@ jobs:
...
+      if: failure()
+      run: |
+        ./utils/test/collect_failed_output.sh ${BUILD_DIR} ${BUILD_DIR}/failed-tests.${TEST_UUID}.tar
+
+    - name: Upload archive of failed test output
```
Comment (bhourahine) [inline .github/workflows/build.yml:141]:
```suggestion
    - name: Upload the archive of failed test output
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -130,9 +130,21 @@ jobs:
...
+      if: failure()
+      uses: actions/upload-artifact@v4
+      with:
+        name: failed-tests.${{ env.TEST_UUID }}.tar
+        path: ${{ env.BUILD_DIR }}/failed-tests.${{ env.TEST_UUID }}.tar
+        retention-days: 3
```
Comment (bhourahine) [inline .github/workflows/build.yml:147]: Would 2 be OK ? (thinking about iterative attempts to fix something that failed, which could end up archiving a lot).

File: utils/test/collect_failed_output.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+#
+# Collect the output of failed tests into a tarball
+#
+# Usage: collect_failed_output.sh <build_dir> <archive_name>
+#
+# build_dir: The build directory where code was built and tested with CMake
+# archive_name: The name of the tarball to create
+#
+# The program assumes that the test names in the failure log are of the form *:<test_name>
+# <test_name> is the name of the test directory below <build_dir>/test.
```
Comment (bhourahine) [inline utils/test/collect_failed_output.sh:11]:
```suggestion
# <test_name> is the name of the test directory that is below <build_dir>/test.
```

**Review by bhourahine — COMMENTED** [review]
Looks generally good (see some minor comments), but is the artifact-url exposed somewhere visible on return from upload-artifact ?

**Comment by bhourahine** [comment]
> Looks generally good (see some minor comments), but is the artifact-url exposed somewhere visible on return from upload-artifact ?

Just tried it and now get that the url appears on stdout.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -130,9 +130,21 @@ jobs:
         else
           echo "TEST_OMP_PROCS=2" >> $GITHUB_ENV
         fi
-        pushd ${BUILD_DIR}
-        ctest -j ${TEST_OMP_PROCS} --output-on-failure
-        popd
+        echo "TEST_UUID=$(uuidgen)" >> ${GITHUB_ENV}
+        ctest -j 2 --output-on-failure --test-dir ${BUILD_DIR}
```
[the remark this answers is not in the document]
Comment (aradi) [reply .github/workflows/build.yml:134]: Thanks for spotting it. It is a copy & paste mistake I've made when resolving the conflicts with main.

File: .github/workflows/build.yml (hunk 3 above)
Comment (aradi) [reply .github/workflows/build.yml:147]: Sure, I've changed it.

**Comment by aradi** [comment]
> Looks generally good (see some minor comments), but is the artifact-url exposed somewhere visible on return from upload-artifact ?

Yes, as you've found out, it it only accessible via the link shown in the stdout. Probably github does not create the menu for artifacts, if the job failed (just guessing). But we need stdout anyway to find out the uuid I use to prevent name conflicts...

**Review by bhourahine — APPROVED** [review]
LGTM

**Pull request merged**